src/constraint/CollisionFree.cpp

```diff
@@ -42,16 +43,39 @@ bool CollisionFree::isSatisfied(
         groups.second.get(),
         mCollisionOptions,
         &collisionResult);
+
     if (collision)
+    {
+      if (_outcome != nullptr)
+      {
+        auto collisionOutcome = static_cast<CollisionFreeOutcome*>(_outcome);
```

Comment:
It's not guaranteed if `_outcome` is the type of `CollisionFreeOutcome*`. I think we should do one of the follows:
(1) use assertion (in debug mode) and `dynamic_cast` to check if `_outcome` is the right type. The downside would be that we still have problem in release mode
(2) run the routine only when it succeded to cast to the right type using `dynamic_cast` (warnings if the type is not the right one). The downside would be the overhaed of using `dynamic_cast`.
(3) throw exception when the type is not right. The downside would be that it has to use `dynamic_cast` anyway.

I'm not sure what would be the preferred way here. @brianhou , @mkoval, @gilwoolee thoughts?

Reply:
Is the only difference between (2) and (3) whether we silently fail or raise an exception when the `TestableOutcome` is the wrong type? In that case, I think I'd prefer (3) over (2) to make sure we know that we've made a mistake.

My (limited) understanding is that we won't be passing a non-null `TestableOutcome` that often, so I'm also not as concerned about the overhead of using a `dynamic_cast` in release mode. Perhaps we can use (3) now and if it becomes performance-critical we can re-evaluate and perhaps switch to (1).

Reply:
I agree with this logic (assuming that @brianhou's reasoning about (2) and (3) is correct). I'm going to with (3) right now, and we can change if people have strong feelings towards either approach.